docs/sources/developers/plugins/build-a-streaming-data-source.md

```diff
@@ -0,0 +1,128 @@
+
+When monitoring critical applications, you want your dashboard to refresh as soon as your data does. In Grafana, you can set your dashboards to automatically refresh at a certain interval, no matter what data source you use. Unfortunately, this means that your queries are requesting all the data to be sent again, regardless of whether the data has actually changed.
+
+By enabling _streaming_ for your data source plugin, you can update your dashboard as soon as new data becomes available.
+
+For example, a streaming data source plugin can subscribe to a message bus using [MQTT](https://mqtt.org/), or [AMQP](https://www.amqp.org/), and update the visualization whenever a new message is available.
+
+Let's see how you can add streaming to an existing data source!
+
+Grafana uses [RxJS](https://rxjs.dev/) to continuously send data from a data source to a panel visualization. If you want to learn more, check out the [RxJS documentation](https://rxjs.dev/guide/overview).
+
+1. Enable streaming for your data source in the `plugin.json` file.
+
+   ```json
+   {
+     "streaming": true
+   }
+   ```
+
+1. Change the signature of the `query` method to return an `Observable` from the `rxjs` package.
+
+   ```ts
+   import { Observable } from 'rxjs';
+   ```
+
+   ```ts
+   query(options: DataQueryRequest<MyQuery>): Observable<DataQueryResponse> {
+     // ...
+   }
+   ```
+
+   Note that you also need to remove the `async` keyword from the method signature.
+
+1. Create an `Observable` for each query and combine them all using the `merge` function from the `rxjs` package.
+
+   ```ts
+   import { Observable, merge } from 'rxjs';
+   ```
+
+   ```ts
+   const observables = options.targets.map(target => {
+     return new Observable<DataQueryResponse>(subscriber => {
+       // ...
+     });
+   });
+
+   return merge(...observables);
+   ```
+
+1. In the `subscribe` function, create a `CircularDataFrame`.
+
+   ```ts
+   const frame = new CircularDataFrame({
+     append: 'tail',
+     capacity: 1000,
+   });
+
+   frame.refId = query.refId;
+   frame.addField({ name: 'time', type: FieldType.time });
+   frame.addField({ name: 'value', type: FieldType.number });
+   ```
+
+   Circular data frames have a limited capacity. When a circular data frame reaches its capacity, the oldest data point is removed.
+
+1. Add a new data point and use `subscriber.next()` to send the updated data frame.
+
+   ```ts
+   const intervalId = setInterval(() => {
+     frame.add({ time: Date.now(), value: Math.random() });
+
+     subscriber.next({
+       data: [frame],
+       key: query.refId,
+     });
+   }, 500);
+
+   return () => {
+     clearInterval(intervalId);
+   };
+   ```
+
+   In this example, we're appending new data every 500 milliseconds. If your data source uses a message bus, you can call `subscriber.next()` whenever you receive a new message.
```

Comment:
The whole point of subscribing to an Observable is that you don't need to pull it with `setInterval`, data will just arrive in the Observable when it arrives :): I usually think of an Observable as an infinite array of data where each index represents a point in time.  Also RxJs has a very steep learning curve so that should also be mentioned somewhere in these docs as this is pretty advanced topic.

Reply:
I'm using `setInterval` here to let the reader try it out without having a message bus available. Also added the note that you can call `next` whenever you receive a new message.

What would be a better example that more accurately describes the intended use case?

Reply:
My bad, it wasn't clear to me we were simulating streaming in the data source. I think this looks ok from that perspective.